rHEALPix projection

pelson opened the pull request:
Relates to #882. (picture available there)

QuLogic commented on lib/cartopy/_crs.pyx:
```diff
@@ -282,7 +282,7 @@ cdef class CRS:
-        if trap and status == -14 or status == -20:
+        if trap and status in [-14, -15, -20]:
             # -14 => "latitude or longitude exceeded limits"
             # -20 => "tolerance condition error"
```
Add comment for -15, too.

QuLogic commented on lib/cartopy/_trace.cpp:
```diff
@@ -63,7 +63,7 @@ Point CartesianInterpolator::project(const Point &src_xy)
+    if (status == -14 || status == -20 || status == -15)
     {
         // -14 => "latitude or longitude exceeded limits"
         // -20 => "tolerance condition error"
```
Add comment for -15 too.

QuLogic commented on lib/cartopy/tests/test_crs.py:
```diff
@@ -225,6 +225,24 @@ def test_utm(self):
 
+def test_rHEALPix_defaults():
+    crs = ccrs.rHEALPix()
+    assert sorted(crs.proj4_params.items()) == [('ellps', 'WGS84'),
```
You can just assert a dictionary equality.

QuLogic commented:
Seems to be a bug in the south square in the docs.

pelson commented:
> Seems to be a bug in the south square in the docs.

https://25-5282596-gh.circle-artifacts.com/0/root/project/docs/build/html/crs/projections.html#rhealpix

![The bug](https://25-5282596-gh.circle-artifacts.com/0/root/project/docs/build/html/_images/projections-27.png)

Seem so. I'll try digging a little - strange as it works for me.

pelson commented:
The docs don't even build for python2 ....

```
reading sources... [ 17%] crs/projections
/root/project/lib/cartopy/mpl/feature_artist.py:136: UserWarning: Unable to determine extent. Defaulting to global.
  warnings.warn('Unable to determine extent. Defaulting to global.')
No handlers could be found for logger "shapely.geos"
Makefile:47: recipe for target 'html' failed
make: *** [html] Segmentation fault
Exited with code 2
```

pelson commented:
I'm definitely seeing the issue with the south pole box reproducibly on my machine now. Any point I hover over in the south pole box tells me it is 90N, so clearly a coordinate transformation issue. My problem has been diagnosing - I'm now struggling to actually install a working version so that I can start bisecting the issue to find the root cause.

For that reason, let's put this PR on hold until I can realistically pick it back up again. (I may close the PR if that gets too long)

pelson commented:
I've just updated this PR on the back of fixing the bug introduced to Proj.4. Images, and the fix I applied to Proj.4 can be found at https://github.com/OSGeo/proj.4/pull/1207#issuecomment-449696833.

The update therefore includes both the Healpix and the RectangularHealpix projections. As far as I could tell, cartopy won't actually compile against v4.9.0 of Proj (it has to be 4.9.1, I think) - I've also added the flag that we should continue to use the deprecated proj API for its next release (v6). #1140 is the issue tracking the update to the new API.

greglucas commented:
I'm going to close this, as there were other changes needed to make this work in newer versions of Cartopy. I've opened up a new PR #2458 with these two projections added.
